File: Java/프로그래머스/2/159993. 미로 탈출/미로 탈출.java
import java.util.ArrayDeque;

class Solution {
    static final int[] dx = {1, -1, 0, 0};
    static final int[] dy = {0, 0, 1, -1};
    
    static class Point {
        int nx, ny;
        
        Point(int nx, int ny) {
            this.nx = nx;
            this.ny = ny;
        }
    }
    
    static char[][] map;
    
    static int r, c;
    
    public int solution(String[] maps) {
        c = maps.length;
        r = maps[0].length();
        
        map = new char[c][r];
        
        
        for (int y = 0; y < c; y++) {
            map[y] = maps[y].toCharArray();
        }
        
        Point start = null, passingPoint = null, end = null;
        
        for (int y = 0; y < c; y++) {
            for (int x = 0; x < r; x++) {
                if (map[y][x] == 'S') {
                    start = new Point(x, y);
                } else if (map[y][x] == 'L') {
                    passingPoint = new Point(x, y);
                } else if (map[y][x] == 'E') {
                    end = new Point(x, y);
                }
            }
        }
        
        int sTopp = bfs(start, passingPoint);
        int ppToe = bfs(passingPoint, end);
        
        if (sTopp == -1 || ppToe == -1) {
            return -1;
        }
        
        return sTopp + ppToe;
    }
    
    static int bfs(Point s, Point d) {
        int[][] distance = new int[c][r];
        ArrayDeque<Point> queue = new ArrayDeque<>();
        distance[s.ny][s.nx] = 1;
        queue.add(new Point(s.nx, s.ny));
        
        while (!queue.isEmpty()) {
            Point cureent = queue.poll();
            for (int i = 0; i < dx.length; i++) {
                int moveY = cureent.ny+dy[i];
                int moveX = cureent.nx+dx[i];
                
                if (moveY < 0 || moveY >= c || moveX < 0 || moveX >= r)
                    continue;
            
                if ((map[moveY][moveX] != 'X') && (distance[moveY][moveX] <= 0)) {
                    distance[moveY][moveX] = distance[cureent.ny][cureent.nx] + 1;
                    queue.add(new Point(moveX, moveY));
                }
                
                if (moveY == d.ny && moveX == d.nx) {
                    return distance[moveY][moveX] - 1;
                }
            }
        }
        
        return -1;
    }
    
    
}

File: Swift/프로그래머스/2/159993. 미로 탈출/미로 탈출.swift
import Foundation

struct Point {
    let x: Int
    let y: Int
    
    init(_ x: Int, _ y:Int) {
        self.x = x
        self.y = y
    }
}

var map: [[String]] = []
let move: [(Int, Int)] = [(1, 0), (-1, 0), (0, 1), (0, -1)]
var xCount: Int = 0
var yCount: Int = 0

func solution(_ maps:[String]) -> Int {
    yCount = maps.count
    xCount = maps.first!.count
    
    for mapString in maps {
        map.append(mapString.map{ String($0) })
    }
    
    var start: Point = Point(0, 0)
    var lever: Point = Point(0, 0)
    var end: Point = Point(0, 0)
    
    
    for y in 0..<yCount {
        for x in 0..<xCount {
            let road = map[y][x]
            if road == "S" {
                start = Point(x, y)
            } else if road == "L" {
                lever = Point(x, y)
            } else if road == "E" {
                end = Point(x, y)
            }
        }
    }
    
    let stol = bfs(start, lever)
    let ltoe = bfs(lever, end)
    
    guard stol >= 0, ltoe >= 0 else { return -1 }
    
    return stol + ltoe
}

func bfs(_ start: Point, _ end: Point) -> Int {
    var distance: [[Int]] = .init(repeating: .init(repeating: 0, count: xCount), count: yCount)
    var queue: [Point] = []
    queue.append(start)
    distance[start.y][start.x] = 1
    
    while !queue.isEmpty {
        let current = queue.removeFirst()
        for dxy in move {
            let movedY = current.y + dxy.1
            let movedX = current.x + dxy.0
            
            guard movedY >= 0, movedY < yCount, movedX >= 0, movedX < xCount else { continue }
            guard distance[movedY][movedX] <= 0 else { continue }
            let road = map[movedY][movedX]
            if road != "X" {
                queue.append(Point(movedX, movedY))
                distance[movedY][movedX] = distance[current.y][current.x] + 1
            }
            if movedY == end.y && movedX == end.x {
                return distance[movedY][movedX] - 1
            }
        }
    }
    return -1
}